slides/k8s/healthchecks-more.md

```diff
@@ -0,0 +1,345 @@
+## Questions to ask before adding healthchecks
+
+- Do we want liveness, readiness, both?
+
+  (sometimes, we can use the same check, but with different failure thresholds)
+
+- Do we have existing HTTP endpoints that we can use?
+
+- Do we need to add new endpoints, or perhaps use something else?
+
+- Are our healthchecks likely to use resources and/or slow down the app?
+
+- Do they depend on additional services?
+
+  (this can be particularly tricky, see next slide)
+
+---
+
+## Healthchecks and dependencies
+
+- A good healthcheck should always indicate the health of the service itself
+
+- It should not be affected by the state of the service's dependencies
+
+- Example: a web server requiring a database connection to operate
+
+  (make sure that the healthcheck can report "OK" even if the database is down)
```

Comment:
What good is a health check if it tells you things are okay when the database is down?

Reply:
Because if the database is down, we shouldn't restart our 10 database frontends (or remove them from the load balancer rotation). But you're right, I should explain that better. I'll improve the wording, perhaps add an extra slide about it.

Reply:
Voilà, new wording:

  (make sure that the healthcheck can report "OK" even if the database is down;
  because it won't help us to restart the web server if the issue is with the DB!)

Reply:
Okay, I wonder if just saying "healthcheck" has the implication of "a end-to-end healthcheck" of the sort that people would use to toggle parts of a site? I have lots of questions about how this sort of check is useful, though. "The front end of the site works and we'll ignore the rest" is not a comprehensive check of anything, clearly. So I think this might raise more questions than it answers...?

Reply:
Oh, it shouldn't be end-to-end; it should rather be as specific as possible. It should indicate the health of the *container* (not even the pod, but the container within the pod), since it will be used by kubelet to determine "oh, that container is FUBAR, I'll restart it meow!" (liveness) or "oh, that container seems very busy at the moment, I'll mark it as NOT-READY so that it is removed from the load balancing rotation associated with its services" (readiness).

Perhaps I should expand this in the beginning of the chapter on healthchecks?

(This file, `healthchecks-more.md`, is an optional "part 2" that comes just after `healthchecks.md`.)

Reply:
I think it's good thus far. I'll run it and then assess what to change.